docs/async-await-proposal.md

```diff
@@ -0,0 +1,442 @@
+
+  /// Splits a request into words and returns each word in a stream of messages.
+  func expand(
+    request: Echo_EchoRequest,
+    responseStreamWriter: AsyncResponseStreamWriter<Echo_EchoResponse>,
+    context: AsyncServerCallContext
+  ) async throws
+
+  /// Collects a stream of messages and returns them concatenated when the caller closes.
+  func collect(
+    requests: GRPCAsyncStream<Echo_EchoRequest>,
+    context: AsyncServerCallContext
+  ) async throws -> Echo_EchoResponse
+
+  /// Streams back messages as they are received in an input stream.
+  func update(
+    requests: GRPCAsyncStream<Echo_EchoRequest>,
+    responseStreamWriter: AsyncResponseStreamWriter<Echo_EchoResponse>,
+    context: AsyncServerCallContext
+  ) async throws
+}
+```
+
+Here is an example implementation of the bidirectional streaming `update`
+handler using this new API:
+
+```swift
+public func update(
+  requests: GRPCAsyncStream<Echo_EchoRequest>,
+  responseStreamWriter: AsyncResponseStreamWriter<Echo_EchoResponse>,
+  context: AsyncServerCallContext
+) async throws {
+  var count = 0
+  for try await request in requests {
+    let response = Echo_EchoResponse.with {
+      $0.text = "Swift echo update (\(count)): \(request.text)"
+    }
+    count += 1
+    try await responseStreamWriter.sendResponse(response)
+  }
+}
+```
+
+This API addresses the previously noted drawbacks the existing API:
+
+> 1. The fact that gRPC is implemented on top of NIO is not hidden from the user
+>   and they need to implement an API in terms of an `EventLoopFuture` and needs
+>   to access an `EventLoop` from the call context.
+
+There is no longer a need for the adopter to `import NIO` nor implement anything
+in terms of NIO types. Instead they now implement an `async` function.
+
+> 2. There is a different context type passed to the user function for each
+>   different type of call and this context is generic over the response type.
+
+The same non-generic `AsyncServerCallContext` is passed to the user function
+regardless of the type of RPC.
+
+> 3. In the server- and bidirectional-streaming call handlers, an added layer of
+>   asynchrony is exposed. That is, the user must return a _future_ for
+>   a closure that will handle incoming events.
+
+The user function consumes requests from an `AsyncSequence`, using the new
+language idioms.
+
+> 4. The user _must_ fulfil the `statusPromise` when it receives `.end` but there
+>   is nothing that enforces this.
+
+The user need simply return from the function or throw an error. The closing of
+the call is handled by the library.
+
+If the user function throws a `GRPCStatus` (which already conforms to `Error`)
+or a value of a type that conforms to `GRPCStatusTransformable` then the library
+will take care of setting the RPC status appropriately. If the user throws
+anything else then the library will still take care of setting the status
+appropriately, but in this case it will use `internalError` for the RPC status.
```

Comment:
nitpick, perhaps `GRPCStatusError` if it is intended to be thrown...? 🤔

Reply:
That's a fair callout. `GRPCStatus` already exists in the API and already conforms to `Error` so I was working with that. It isn't only used on the failure path which is why I guess it does not have `Error` in the name.

I suppose we could wrap the `GRPCStatus` in a struct with a more explicit name. @glbrntt do you have any thoughts on this?

Reply:
I agree `GRPCStatusError` is a better name (and shouldn't allow an 'ok' status) but personally I don't think we buy ourselves much by wrapping a `GRPCStatus`. 

The existing name would still be present in various places throughout the API (we could deprecate it but it's quite pervasive so I'd rather not) and having two very similar types would make the overall picture less clear, I think.

We should revisit when we do a major breaking change though.